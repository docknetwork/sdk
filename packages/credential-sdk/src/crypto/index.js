export * from '@docknetwork/crypto-wasm-ts/lib/index.js';
