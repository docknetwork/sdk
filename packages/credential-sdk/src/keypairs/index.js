export { default as DockKeypair } from './keypair';
export { default as Ed25519Keypair } from './keypair-ed25519';
export { default as Secp256k1Keypair } from './keypair-secp256k1';
export { default as DidKeypair } from './did-keypair';
