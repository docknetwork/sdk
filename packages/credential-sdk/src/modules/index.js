export * from './abstract';
export * from './multi-api';
