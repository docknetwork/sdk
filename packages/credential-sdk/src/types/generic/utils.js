import TypedBytes from './typed-bytes';
import TypedArray from './typed-array';

export class ByteArray extends TypedBytes {}

export class ArrayOfByteArrays extends TypedArray {
  static Class = TypedBytes;
}
