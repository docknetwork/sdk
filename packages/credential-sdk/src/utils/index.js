export * from './assert';
export * from './generic';
export * from './interfaces';
export * from './ss58';
export * from './async';
export * from './json-fetch';
export * from './type-helpers';
export * from './bytes';
export * from './inheritance';
export * from './misc';
