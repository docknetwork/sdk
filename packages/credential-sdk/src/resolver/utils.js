import { fmtIter } from '../utils/generic';

/**
 * Caches last function result and returns it if function called with the same args again.
 * @param {Function}
 * @returns {Function}
 */
export const cacheLast = (fn) => {
  let lastArgs;
  let lastValue;

  return function cached(...args) {
    if (lastArgs !== void 0) {
      let same = true;
      const length = Math.max(lastArgs.length, args.length);
      for (let i = 0; i < length && same; ++i) {
        same &&= lastArgs[i] === args[i];
      }

      if (same) {
        return lastValue;
      }
    }
    lastValue = fn.apply(this, args);
    lastArgs = args;

    return lastValue;
  };
};

/**
 * Ensures that each item is present in the allowed set and that both sets are valid.
 *
 * @template I
 * @template W
 * @param {Iterable<I | W>} items
 * @param {Iterable<T | W>} allowed
 * @param {W} wildcard
 */
export const ensureItemsAllowed = (items, allowed, wildcard) => {
  const itemsSet = new Set(items);
  const allowedSet = new Set(allowed);
  if (items.size === 0) {
    throw new Error('Empty item set provided');
  }
  if (allowed.size === 0) {
    throw new Error('Empty allowed set provided');
  }

  for (const toCheck of [itemsSet, allowedSet]) {
    if (toCheck.has(wildcard) && toCheck.size > 1) {
      throw new Error(
        `Can't have wildcard mixed with other patterns for ${fmtIter(toCheck)}`,
      );
    }
  }

  if (allowedSet.has(wildcard)) return;

  for (const item of itemsSet) {
    if (!allowedSet.has(item)) {
      throw new Error(
        `Item not found in ${fmtIter(allowedSet)}: \`${item.toString()}\``,
      );
    }
  }
};

export const itemsOrWildcard = (items, wildcard) => {
  for (const item of items) {
    if (item === wildcard) {
      return wildcard;
    }
  }

  return [...new Set(items)];
};
