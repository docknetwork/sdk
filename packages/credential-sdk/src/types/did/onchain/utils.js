/* eslint-disable max-classes-per-file */

import { isHexWithGivenByteSize } from '../../../utils/bytes';
import {
  PublicKey, // eslint-disable-line
} from '../../public-keys';

import { Signature } from "../../signatures"; // eslint-disable-line

import { DockDIDByteSize } from './constants';

/**
 * Check if the given identifier is the hex representation of a Dock DID.
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDHexIdentifier(identifier) {
  // Byte size of the Dock DID identifier, i.e. the `DockDIDQualifier` is not counted.
  if (!isHexWithGivenByteSize(identifier, DockDIDByteSize)) {
    throw new Error(`DID identifier must be ${DockDIDByteSize} bytes`);
  }
}

/**
 * Check if the given identifier is 32 byte valid SS58 string
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDSS58Identifier(identifier) {
  // base58-check regex
  const regex = new RegExp(/^[5KL][1-9A-HJ-NP-Za-km-z]{47}$/);
  const matches = regex.exec(identifier);
  if (!matches) {
    throw new Error('The identifier must be 32 bytes and valid SS58 string');
  }
}

/**
 * Returns a `DidKey` as expected by the Substrate node
 * @param {PublicKey} publicKey - The public key for the DID. The Keyring is intentionally avoided here as it may not be
 * accessible always, like in case of hardware wallet
 * @param {VerificationRelationship} verRel
 * @returns {object} - The object has structure and keys with same names as expected by the Substrate node
 */
export function createDidKey(publicKey, verRel) {
  return {
    publicKey: publicKey.toJSON(),
    verRels: verRel.value,
  };
}

/**
 *
 * @param {string|DockDidOrDidMethodKey} did - DID as string or an object
 * @param {number} keyId -
 * @param rawSig
 * @param {Signature} sig
 * @returns {object}
 */
export function createDidSig(did, { keyId }, rawSig) {
  const sig = rawSig.toJSON();

  if (did.isDid) {
    return {
      DidSignature: {
        did: did.asDid,
        keyId,
        sig,
      },
    };
  } else if (did.isDidMethodKey) {
    return {
      DidMethodKeySignature: {
        didMethodKey: did.asDidMethodKey,
        sig,
      },
    };
  } else {
    throw new Error(
      `Incorrect DID passed: \`${did}\`, expected instance of either \`DockDid\` or \`DidMethodKey\``,
    );
  }
}
