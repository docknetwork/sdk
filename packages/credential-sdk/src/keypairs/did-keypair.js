import DockKeypair from './keypair';
import { DidKey } from '../types/did/onchain/did-key';
import VerificationMethodRefOrIdentRef from '../types/did/document/verification-method-ref-or-ident-ref';
import { ensureInstanceOf } from '../utils';
import { DidMethodKey } from '../types/did/onchain/typed-did';

/**
 * A class representation of DID Keypair which includes a signing keypair and its associated verification method reference identifier.
 */
export default class DidKeypair {
  /**
   * Constructs a `DidKeypair` instance from the supplied keypair.
   *
   * @param {VerificationMethodRefOrIdentRef} keyRef - The reference of the verification method.
   * @param {DockKeypair} keyPair - Key pair provided.
   */
  constructor(keyRef, keyPair) {
    const ref = VerificationMethodRefOrIdentRef.from(keyRef);

    ensureInstanceOf(keyPair, DockKeypair);

    this.verificationMethodId = ref;
    this.keyPair = keyPair;
  }

  /**
   * Static function to create a `DidKeypair` instance from the provided keyPair.
   *
   * @param {DockKeypair} keyPair - Key pair provided.
   * @return The new instance of `DidKeypair`.
   */
  static didMethodKey(keyPair) {
    const did = DidMethodKey.fromKeypair(keyPair);

    return new this([did, did.asDidMethodKey.toEncodedString()], keyPair);
  }

  /**
   * Get the public key from the current key pair.
   *
   * @return The public key.
   */
  publicKey() {
    return this.keyPair.publicKey();
  }

  /**
   * Creates a new `DidKey` instance from the public key of the current key pair.
   *
   * @param {?number} verRels - Number indicating verification relationships.
   * @return The new instance of `DidKey`.
   */
  didKey(verRels) {
    return new DidKey(this.keyPair.publicKey(), verRels);
  }

  /**
   * Get the DID from the current verification method ID.
   *
   * @return The DID.
   */
  get did() {
    return this.verificationMethodId[0];
  }

  /**
   * Get the Key ID from the current verification method ID.
   *
   * @return The Key ID.
   */
  get keyId() {
    return this.verificationMethodId[1];
  }

  /**
   * Sign the provided message with the key pair.
   *
   * @param {Uint8Array} message - The message to sign.
   * @return The signature.
   */
  sign(message) {
    return this.keyPair.sign(message);
  }
}
