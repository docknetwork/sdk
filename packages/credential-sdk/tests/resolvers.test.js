import { Ed25519Keypair } from "../src/keypairs";
import { AbstractDIDModule } from "../src/modules";
import AbstractApiProvider from "../src/modules/abstract/common/abstract-api-provider";
import {
  Resolver,
  ResolverRouter,
  DIDResolverWithDIDReplacement,
  WILDCARD,
  WildcardResolverRouter,
  createResolver,
  DIDResolver,
} from "../src/resolver";
import {
  CheqdTestnetDIDDocument,
  DockDid,
  DIDDocument,
  DidKey,
} from "../src/types";
import { decodeFromBase58 } from "../src/utils";

class APrefixBMethodResolver extends Resolver {
  prefix = "a";
  method = "b";

  async resolve(id) {
    return this.supports(id) ? `ab-full-${id}` : null;
  }
}

class CPrefixDMethodResolver extends Resolver {
  prefix = "c";
  method = "d";

  async resolve(id) {
    return this.supports(id) ? `cd-full-${id}` : null;
  }
}

class BMethod extends ResolverRouter {
  method = "b";
}

class APrefix extends ResolverRouter {
  prefix = "a";
}

class APrefixBMethod extends ResolverRouter {
  method = "b";
}

class APrefixBMethodResolverExtended extends Resolver {
  prefix = "a";
  method = "b";

  async resolve(id) {
    return this.supports(id) ? `ab-extended-${id}` : null;
  }
}

class APrefixCDMethod extends Resolver {
  prefix = "a";
  method = ["c", "d"];

  async resolve(id) {
    return this.supports(id) ? `a-cd-${id}` : null;
  }
}

class APrefixWildcardMethodResolver extends Resolver {
  prefix = "a";
  method = WILDCARD;

  async resolve(id) {
    return this.supports(id) ? `a-wildcard-${id}` : null;
  }
}

class WildcardPrefixBMethodResolver extends Resolver {
  prefix = WILDCARD;
  method = "b";

  async resolve(id) {
    return this.supports(id) ? `wildcard-b-${id}` : null;
  }
}

class WildcardPrefixAndMethodResolver extends Resolver {
  prefix = WILDCARD;
  method = WILDCARD;

  async resolve(id) {
    return this.supports(id) ? `wildcard-wildcard-${id}` : null;
  }
}

describe("Resolvers", () => {
  it("checks `DIDResolverWithDIDReplacement`", async () => {
    const did = "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5";
    const document = DIDDocument.create(
      did,
      [
        new DidKey({
          ed25519: decodeFromBase58(
            "4qGHUs2Lofp8YH7EtkfUBxcA2B9A3KNHfkhwfd4mC4ZS"
          ),
        }),
      ],
      [
        did,
        "did:cheqd:testnet:ec2ffc8f-bef0-4307-b1dc-1e945aa05a04",
        "did:cheqd:testnet:13426782-8e61-4ea3-9255-20de06ef13ba",
      ]
    );

    const dummyApi = new (class Module extends AbstractDIDModule {
      methods() {
        return ["dock"];
      }

      async getDocument(_) {
        return document.toCheqd(
          CheqdTestnetDIDDocument,
          "0f6edb02-59d7-4cc2-aabd-dd205badb2d5"
        );
      }

      createDocumentTx() {}

      updateDocumentTx() {}

      removeDocumentTx() {}
    })(
      new (class Api extends AbstractApiProvider {
        methods() {
          return ["dock"];
        }
        isInitialized() {
          return true;
        }
        supportsIdentifier() {
          return true;
        }
        stateChangeBytes() {}
        signAndSend() {}
      })()
    );

    expect(
      await new DIDResolverWithDIDReplacement(dummyApi).resolve(String(did))
    ).toMatchSnapshot();
    expect(
      await new DIDResolver(dummyApi).resolve(String(did))
    ).toMatchSnapshot();
  });

  it("checks `ResolverRouter`", async () => {
    expect(() => new ResolverRouter()).toThrowError(
      "No resolvers were provided. You need to either implement `resolve` or provide a list of resolvers"
    );
    expect(() => new BMethod()).toThrowError(
      "No resolvers were provided. You need to either implement `resolve` or provide a list of resolvers"
    );
    expect(() => new APrefixBMethod()).toThrowError(
      "No resolvers were provided. You need to either implement `resolve` or provide a list of resolvers"
    );
    expect(() => new APrefix()).toThrowError(
      "No resolvers were provided. You need to either implement `resolve` or provide a list of resolvers"
    );
    expect(await new APrefixBMethodResolver().resolve("a:b:123")).toBe(
      "ab-full-a:b:123"
    );
    expect(new APrefixBMethodResolver().supports("a:b:123")).toBe(true);
    expect(new APrefixBMethodResolver().supports("a:c:123")).toBe(false);
    expect(new APrefixBMethodResolver().supports("c:b:123")).toBe(false);
    expect(new APrefixBMethodResolver().supports("a:b:123")).toBe(true);
    expect(new APrefixBMethodResolver().supports("a:c:123")).toBe(false);
    expect(new APrefixBMethodResolver().supports("c:b:123")).toBe(false);

    expect(await new APrefixBMethodResolverExtended().resolve("a:b:456")).toBe(
      "ab-extended-a:b:456"
    );
    expect(
      await new APrefixBMethod([new APrefixBMethodResolver()]).resolve(
        "a:b:456"
      )
    ).toBe("ab-full-a:b:456");
    expect(
      await new APrefixBMethod([new APrefixBMethodResolverExtended()]).resolve(
        "a:b:456"
      )
    ).toBe("ab-extended-a:b:456");

    const wildcard = new WildcardResolverRouter([
      new APrefixBMethodResolver(),
      new CPrefixDMethodResolver(),
      new APrefixWildcardMethodResolver(),
      new WildcardPrefixBMethodResolver(),
      new WildcardPrefixAndMethodResolver(),
    ]);

    expect(await wildcard.resolve("a:b:")).toBe("ab-full-a:b:");
    expect(await wildcard.resolve("a:asdasdas:")).toBe(
      "a-wildcard-a:asdasdas:"
    );
    expect(await wildcard.resolve("asdasdasd:b:")).toBe(
      "wildcard-b-asdasdasd:b:"
    );
    expect(await wildcard.resolve("c:d:")).toBe("cd-full-c:d:");
    expect(await wildcard.resolve("asdasdasd:asdasdasd:")).toBe(
      "wildcard-wildcard-asdasdasd:asdasdasd:"
    );

    const nestedWildcard = new WildcardResolverRouter([
      new APrefixBMethodResolverExtended(),
      new WildcardResolverRouter([wildcard]),
    ]);
    expect(await nestedWildcard.resolve("a:b:")).toBe("ab-extended-a:b:");
    expect(await nestedWildcard.resolve("c:d:")).toBe("cd-full-c:d:");
    expect(await wildcard.resolve("a:asdasdas:")).toBe(
      "a-wildcard-a:asdasdas:"
    );
    expect(await nestedWildcard.resolve("asdasdasd:asdasdasd:")).toBe(
      "wildcard-wildcard-asdasdasd:asdasdasd:"
    );

    const abTop = new APrefixBMethodResolverExtended();
    const ab = new APrefixBMethodResolver();
    const cd = new CPrefixDMethodResolver();
    const acdTop = new APrefixCDMethod();
    const awildcard = new APrefixWildcardMethodResolver();
    const widlcardb = new WildcardPrefixBMethodResolver();

    const discreteWrappedInWildcard = new WildcardResolverRouter([
      new WildcardResolverRouter([ab, cd, awildcard, widlcardb]),
      acdTop,
      abTop,
    ]);

    expect(discreteWrappedInWildcard.matchingResolver("a:b:")).toBe(abTop);
    expect(
      discreteWrappedInWildcard
        .matchingResolver("c:d:")
        .matchingResolver("c:d:")
    ).toBe(cd);
    expect(discreteWrappedInWildcard.matchingResolver("a:c:")).toBe(acdTop);
    expect(discreteWrappedInWildcard.matchingResolver("a:d:")).toBe(acdTop);
    expect(
      discreteWrappedInWildcard
        .matchingResolver("a:e:")
        .matchingResolver("a:e:")
    ).toBe(awildcard);
    expect(discreteWrappedInWildcard.matchingResolver("c:f:")).toBe(null);
    expect(
      discreteWrappedInWildcard
        .matchingResolver("x:b:")
        .matchingResolver("x:b:")
    ).toBe(widlcardb);
    expect(discreteWrappedInWildcard.matchingResolver("e:a:")).toBe(null);
  });

  it(`Initializes \`ResolverRouter\` with duplicate items`, () => {
    new ResolverRouter([
      new (class extends Resolver {
        prefix = ["b", "b"];
        method = ["a", "a"];
      })(),
    ]);
  });

  it("checks `createResolver`", async () => {
    const resolve = async () => 1;

    expect(() =>
      createResolver(new APrefixBMethodResolver(), { prefix: "c" })
    ).toThrowError("Item not found in [c]: `a`");

    expect(() =>
      createResolver(new APrefixBMethodResolver(), { method: "c" })
    ).toThrowError("Item not found in [c]: `b`");

    const singleResolver = createResolver(resolve, {
      prefix: "abc",
      method: "cde",
    });

    expect(await singleResolver.resolve("abc:cde:1")).toBe(1);
    expect(singleResolver.supports("abc:cde:1")).toBe(true);
    expect(singleResolver.supports("abc:de:1")).toBe(false);

    const wildcardResolver = createResolver(
      createResolver(resolve, { prefix: "abc", method: "cde" })
    );

    expect(await wildcardResolver.resolve("abc:cde:1")).toBe(1);
    expect(wildcardResolver.supports("abc:cde:1")).toBe(true);
    expect(wildcardResolver.supports("abc:de:1")).toBe(true);

    const multiResolver = createResolver(resolve, {
      prefix: ["abc"],
      method: ["cde"],
    });
    expect(await multiResolver.resolve("abc:cde:1")).toBe(1);
    expect(multiResolver.supports("abc:cde:1")).toBe(true);
    expect(multiResolver.supports("abc:de:1")).toBe(false);
  });
});
