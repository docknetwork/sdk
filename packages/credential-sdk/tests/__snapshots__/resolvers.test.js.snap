// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers checks \`DIDResolverWithDIDReplacement\` 1`] = `
{
  "alsoKnownAs": [],
  "assertionMethod": [
    "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5#keys-1",
  ],
  "authentication": [
    "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5#keys-1",
  ],
  "capabilityDelegation": [],
  "capabilityInvocation": [
    "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5#keys-1",
  ],
  "context": [
    "https://www.w3.org/ns/did/v1",
  ],
  "controller": [
    "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5",
    "did:cheqd:testnet:ec2ffc8f-bef0-4307-b1dc-1e945aa05a04",
    "did:cheqd:testnet:13426782-8e61-4ea3-9255-20de06ef13ba",
  ],
  "id": "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5",
  "keyAgreement": [],
  "service": [],
  "verificationMethod": [
    {
      "controller": "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5",
      "id": "did:dock:5EbpmcZhMPPLCyP4mDwo4bNtwZBi3dZuKzz65PGk2Amnvek5#keys-1",
      "verificationMaterial": "4qGHUs2Lofp8YH7EtkfUBxcA2B9A3KNHfkhwfd4mC4ZS",
      "verificationMethodType": "Ed25519VerificationKey2018",
    },
  ],
  "versionId": "0f6edb02-59d7-4cc2-aabd-dd205badb2d5",
}
`;

exports[`Resolvers checks \`DIDResolverWithDIDReplacement\` 2`] = `
{
  "alsoKnownAs": [],
  "assertionMethod": [
    "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb#keys-1",
  ],
  "authentication": [
    "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb#keys-1",
  ],
  "capabilityDelegation": [],
  "capabilityInvocation": [
    "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb#keys-1",
  ],
  "context": [
    "https://www.w3.org/ns/did/v1",
  ],
  "controller": [
    "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb",
    "did:cheqd:testnet:ec2ffc8f-bef0-4307-b1dc-1e945aa05a04",
    "did:cheqd:testnet:13426782-8e61-4ea3-9255-20de06ef13ba",
  ],
  "id": "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb",
  "keyAgreement": [],
  "service": [],
  "verificationMethod": [
    {
      "controller": "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb",
      "id": "did:cheqd:testnet:b05a3c71-113a-4949-a7ea-3e57d31fd6fb#keys-1",
      "verificationMaterial": "4qGHUs2Lofp8YH7EtkfUBxcA2B9A3KNHfkhwfd4mC4ZS",
      "verificationMethodType": "Ed25519VerificationKey2018",
    },
  ],
  "versionId": "0f6edb02-59d7-4cc2-aabd-dd205badb2d5",
}
`;
